[CORE] Fold engine into IndexShard

s1monw opened the pull request:
This commit removes most of the Engine abstractions and removes
Engine exposure via dependency injection. It also removes the Holder
abstraction and makes the engine itself start at constrcution time.
It removes the start method from the engine entire which means no engine
instances exists until it's started. There is also no way to stop the
engine to restart, it needs to be an entire new Engine

dakrone commented on src/main/java/org/elasticsearch/index/shard/AbstractIndexShardComponent.java:
```diff
@@ -23,6 +23,7 @@
 import org.elasticsearch.common.logging.Loggers;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.indices.cluster.IndicesClusterStateService;
```
Extra unneeded additions in this file?

mikemccand commented:
LGTM, this is a nice simplification!

dakrone commented on src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -0,0 +1,305 @@
+    private long gcDeletesInMillis = DEFAULT_GC_DELETES.millis();
+
+    private volatile boolean enableGcDeletes = true;
+    private volatile String codecName = DEFAUTL_CODEC;
```
DEFAUTL -> DEFAULT

dakrone commented on src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -0,0 +1,305 @@
+    private final Engine.FailedEngineListener failedEngineListener;
+
+
+    public static final String INDEX_INDEX_CONCURRENCY = "index.index_concurrency";
```
Can probably just call this `INDEX_CONCURRENCY` instead of `INDEX_INDEX_CONCURRENCY`

dakrone commented on src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -0,0 +1,305 @@
+    public static final String INDEX_CODEC = "index.codec";
+    public static final TimeValue DEFAULT_REFRESH_INTERVAL = new TimeValue(1, TimeUnit.SECONDS);
+    public static final TimeValue DEFAULT_GC_DELETES = TimeValue.timeValueSeconds(60);
+    public static final String DEFAUTL_CODEC = "default";
```
DEFAUTL -> DEFAULT again

dakrone commented on src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -0,0 +1,305 @@
+        setCompoundOnFlush(indexSettings.getAsBoolean(EngineConfig.INDEX_COMPOUND_ON_FLUSH, compoundOnFlush));
+        setIndexConcurrency(indexSettings.getAsInt(EngineConfig.INDEX_INDEX_CONCURRENCY, Math.max(IndexWriterConfig.DEFAULT_MAX_THREAD_STATES, (int) (EsExecutors.boundedNumberOfProcessors(indexSettings) * 0.65))));
+        setCodecName(indexSettings.get(EngineConfig.INDEX_CODEC, EngineConfig.DEFAUTL_CODEC));
+        setIndexingBufferSize(indexSettings.getAsBytesSize("index.shard.index_buffer_size", new ByteSizeValue(64, ByteSizeUnit.MB)));
```
This string could be a `static final String` also I think

dakrone replied:
As well as the default buffer size

rjernst commented on src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java:
```diff
@@ -1778,27 +1734,60 @@ private static SegmentReader segmentReader(LeafReader reader) {
 
-    boolean isFailEngineOnCorruption() {
-        return failEngineOnCorruption;
+    class MergeScheduleListener implements MergeSchedulerProvider.Listener {
```
nit: should this be `Scheduler` instead of `Schedule`?

dakrone commented on src/main/java/org/elasticsearch/index/engine/EngineConfig.java:
```diff
@@ -0,0 +1,305 @@
+            final boolean failOnMergeFailure = settings.getAsBoolean(EngineConfig.INDEX_FAIL_ON_MERGE_FAILURE, config.isFailOnMergeFailure());
+            if (failOnMergeFailure != config.isFailOnMergeFailure()) {
+                logger.info("updating {} from [{}] to [{}]", EngineConfig.INDEX_FAIL_ON_MERGE_FAILURE, config.isFailOnMergeFailure(), failOnMergeFailure);
+                config.setFailOnMergeFailure(failOnMergeFailure);
```
Does this need to set `change = true` also? It's missing from this `if` block

s1monw replied:
done

dakrone commented on src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java:
```diff
@@ -163,58 +158,58 @@
+        this.codecService = config.getCodecService();
         this.versionMap = new LiveVersionMap();
-        this.dirtyLocks = new Object[indexConcurrency * 50]; // we multiply it to have enough...
+        this.dirtyLocks = new Object[config.getIndexConcurrency() * 50]; // we multiply it to have enough...
```
Can the collapse the multiplying by 50 into the setting itself? It feels arbitrary and I have no idea where the 50 came from

s1monw replied:
me neither - the setting is used for something else so I can't fold it in directly. I might create one for this

dakrone commented on src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java:
```diff
@@ -163,58 +158,58 @@
+        this.failedEngineListener = config.getFailedEngineListener();
         throttle = new IndexThrottle();
+        this.config = config;
+        listener = new EngineConfig.ApplySettings(logger, config) {
```
Not sure if this should be a concrete inner class also, to avoid the GC issue you mentioned to me before?

s1monw replied:
afaik the problem only occurs if the ann. inner class is bound in a guice context

dakrone commented on src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java:
```diff
@@ -322,67 +311,31 @@ public void close() throws IOException {
-    }
-
-    @Override
     public ShardId shardId() {
```
I think this can be removed (here and from the interface) and then `this.shardId` can be used in the only caller, the `.get()` method

dakrone commented on src/main/java/org/elasticsearch/index/engine/internal/InternalEngine.java:
```diff
@@ -1468,16 +1424,16 @@ private IndexWriter createWriter() throws IOException {
-            config.setCodec(codecService.codec(codecName));
+            config.setRAMBufferSizeMB(this.config.getIndexingBufferSize().mbFrac());
+            config.setMaxThreadStates(this.config.getIndexConcurrency());
+            config.setCodec(codecService.codec(this.config.getCodecName()));
```
Can we name this `writerConfig`? It's weird to shadow the class `config` since it looks like the same thing (though it isn't)

dakrone commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -874,6 +898,10 @@ private Query filterQueryIfNeeded(Query query, String[] types) {
     public static final String INDEX_REFRESH_INTERVAL = "index.refresh_interval";
 
+    public void addFailedEngineListener(Engine.FailedEngineListener failedEngineListener) {
+        this.failedEngineListener.failedEngineListeners.add(failedEngineListener);
```
Waaaay too many "failedEngineListener"s in this line :) maybe we can rename one/some to make it a little less mind-boggling?

dakrone commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -993,4 +1021,46 @@ private void checkIndex(boolean throwException) throws IndexShardException {
+        }
+    }
+
+    protected Engine newEngineInstance() {
```
Can this just go into `createNewEngine`? It's the only called of this method.

dakrone commented on src/test/java/org/elasticsearch/index/engine/internal/InternalEngineTests.java:
```diff
@@ -231,8 +225,21 @@ protected Engine createEngine(IndexSettingsService indexSettingsService, Store s
+                new AnalysisService(shardId.index(), indexSettingsService.getSettings()), new SimilarityService(shardId.index()), new CodecService(shardId.index()), new Engine.FailedEngineListener() {
+            @Override
+            public void onFailedEngine(ShardId shardId, String reason, @Nullable Throwable t) {
+
```
Can you add a comment why this is a no-op?

dakrone commented:
Left some comments, but I like the simplification a lot!

s1monw commented:
I iterated here a bit and added all kinds of cleanups on the engine. I also added documentation to the new EngineConfig and removed unused methods. I think it's ready, unless anybody objects I will push soon

jpountz commented:
LGTM

s1monw commented:
I pushed this to 1.x as well
